### PR TITLE
V0.12.1.x govobj sync

### PR DESCRIPTION
**Pull request by evan82**
- disable fCached values
- use two maps for storing votes, by hash and parent-hash/type
- disable part of flatdb.dump (still overwriting)
- fixed govobj/votes relay and sync

### CODE REVIEW

File: src/version.h [hunk 1]
```diff
@@ -19,22 +19,22 @@ static const int INIT_PROTO_VERSION = 209;
...
 //! minimum peer version for masternode winner broadcasts
 static const int MIN_MNW_PEER_PROTO_VERSION = 70103;
 
 //! minimum peer version that can receive masternode payments
 // V1 - Last protocol version before update
 // V2 - Newest protocol version
-static const int MIN_MASTERNODE_PAYMENT_PROTO_VERSION_1 = 70066;
-static const int MIN_MASTERNODE_PAYMENT_PROTO_VERSION_2 = 70103;
+static const int MIN_MASTERNODE_PAYMENT_PROTO_VERSION_1 = 70103;
+static const int MIN_MASTERNODE_PAYMENT_PROTO_VERSION_2 = 70200;
```
Comment (schinzelh) [inline src/version.h]: Shouldnt it be '70201'? '70200' was a internal testnet version that will never make it to mainnet.

File: src/version.h [hunk 2]
```diff
@@ -19,22 +19,22 @@ static const int INIT_PROTO_VERSION = 209;
 static const int GETHEADERS_VERSION = 70077;
 
 //! disconnect from peers older than this proto version
-static const int MIN_PEER_PROTO_VERSION = 70066;
+static const int MIN_PEER_PROTO_VERSION = 70103;
 
 //! minimum peer version accepted by DarksendPool
-static const int MIN_POOL_PEER_PROTO_VERSION = 70103;
+static const int MIN_POOL_PEER_PROTO_VERSION = 70200;
```
Comment (schinzelh) [inline src/version.h]: same here

File: src/version.h [hunk 3]
```diff
@@ -19,22 +19,22 @@ static const int INIT_PROTO_VERSION = 209;
...
 
 //! minimum peer version for masternode budgets
-static const int MIN_BUDGET_PEER_PROTO_VERSION = 70103;
+static const int MSG_GOVERNANCE_PEER_PROTO_VERSION = 70200;
```
Comment (schinzelh) [inline src/version.h]: same here

File: src/masternode-sync.h [hunk 4]
```diff
@@ -9,9 +9,9 @@
 #define MASTERNODE_SYNC_SPORKS            1
 #define MASTERNODE_SYNC_LIST              2
 #define MASTERNODE_SYNC_MNW               3
-#define MASTERNODE_SYNC_BUDGET            4
-#define MASTERNODE_SYNC_BUDGET_PROP       10
-#define MASTERNODE_SYNC_BUDGET_FIN        11
+#define MASTERNODE_SYNC_GOVERNANCE            4
+#define MASTERNODE_SYNC_GOVOBJ       10
+#define MASTERNODE_SYNC_GOVERNANCE_FIN        11
 #define MASTERNODE_SYNC_FAILED            998
```
Comment (schinzelh) [inline src/masternode-sync.h]: Can we preserve the formatting? :-)

**Comment by schinzelh** [comment]
some nits but utACK

File: src/governance.cpp [hunk 5]
```diff
@@ -99,6 +97,116 @@ bool IsCollateralValid(uint256 nTxCollateralHash, uint256 nExpectedHash, std::st
...
+        }
+
+        // ask for a specific proposal and it's votes
+        Sync(pfrom, nProp);
+        LogPrint("mngovernance", "syncing governance objects to our peer at %s\n", pfrom->addr.ToString());
+    }
+
+    // NEW GOVERNANCE OBJECT
+
+    if (strCommand == NetMsgType::MNGOVERNANCEOBJECT) {
```
Comment (UdjinM6) [inline src/governance.cpp]: should be `else if` to follow the same style as in other `ProcessMessage`

File: src/governance.cpp [hunk 6]
```diff
@@ -99,6 +97,116 @@ bool IsCollateralValid(uint256 nTxCollateralHash, uint256 nExpectedHash, std::st
...
+        masternodeSync.AddedBudgetItem(govobj.GetHash());
+
+        LogPrintf("Governance object - new! - %s\n", govobj.GetHash().ToString());
+        //We might have active votes for this proposal that are valid now
+        CheckOrphanVotes();
+    }
+
+    // NEW GOVERNANCE OBJECT VOTE
+
+    if (strCommand == NetMsgType::MNGOVERNANCEVOTE) {
```
Comment (UdjinM6) [inline src/governance.cpp]: same here, `else if`

File: src/governance.cpp [hunk 7]
```diff
@@ -99,6 +97,116 @@ bool IsCollateralValid(uint256 nTxCollateralHash, uint256 nExpectedHash, std::st
...
+            mnodeman.AskForMN(pfrom, vote.vinMasternode);
+            return;
+        }
+
+        mapSeenVotes.insert(make_pair(vote.GetHash(), SEEN_OBJECT_IS_VALID));
+        if(!vote.IsValid(true)){
+            LogPrintf("mngovernance - signature invalid\n");
+            if(masternodeSync.IsSynced()) Misbehaving(pfrom->GetId(), 20);
+            // it could just be a non-synced masternode
+            mnodeman.AskForMN(pfrom, vote.vinMasternode);
```
Comment (UdjinM6) [inline src/governance.cpp]: `mapSeenVotes.insert(make_pair(vote.GetHash(), SEEN_OBJECT_ERROR_INVALID));` here?

File: src/governance.cpp [hunk 8]
```diff
@@ -190,7 +299,7 @@ CGovernanceObject *CGovernanceManager::FindGovernanceObject(const std::string &s
     return pGovObj;
 }
 
-CGovernanceObject *CGovernanceManager::FindGovernanceObject(uint256 nHash)
+CGovernanceObject *CGovernanceManager::FindGovernanceObject(uint256& nHash)
```
Comment (UdjinM6) [inline src/governance.cpp]: `uint256& nHash` -> `const uint256& nHash`

File: src/governance.cpp [hunk 9]
```diff
@@ -320,190 +429,54 @@ void CGovernanceManager::NewBlock()
...
         ++it2;
     }
 
-    pfrom->PushMessage(NetMsgType::SYNCSTATUSCOUNT, MASTERNODE_SYNC_BUDGET_PROP, nInvCount);
-
+    pfrom->PushMessage(NetMsgType::SYNCSTATUSCOUNT, MASTERNODE_SYNC_GOVOBJ, nInvCount);
     LogPrintf("CGovernanceManager::Sync - sent %d items\n", nInvCount);
 }
 
+void CGovernanceManager::SyncParentObjectByVote(CNode* pfrom, CGovernanceVote& vote)
```
Comment (UdjinM6) [inline src/governance.cpp]: `CGovernanceVote& vote` -> `const CGovernanceVote& vote`

File: src/governance.h [hunk 10]
```diff
@@ -94,12 +98,13 @@ class CGovernanceManager
...
     //void MarkSynced();
     void Sync(CNode* node, uint256 nProp);
+    void SyncParentObjectByVote(CNode* pfrom, CGovernanceVote& vote);
 
     void ProcessMessage(CNode* pfrom, std::string& strCommand, CDataStream& vRecv);
     void NewBlock();
 
     CGovernanceObject *FindGovernanceObject(const std::string &strName);
-    CGovernanceObject *FindGovernanceObject(uint256 nHash);
+    CGovernanceObject *FindGovernanceObject(uint256& nHash);
```
Comment (UdjinM6) [inline src/governance.h]: const

File: src/governance.h [hunk 11]
```diff
@@ -94,12 +98,13 @@ class CGovernanceManager
     //void ResetSync();
     //void MarkSynced();
     void Sync(CNode* node, uint256 nProp);
+    void SyncParentObjectByVote(CNode* pfrom, CGovernanceVote& vote);
```
Comment (UdjinM6) [inline src/governance.h]: const

File: src/main.cpp [hunk 12]
```diff
@@ -4581,9 +4605,8 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                 }
 
 
-                if (!pushed) {
+                if (!pushed)
```
Comment (UdjinM6) [inline src/main.cpp]: write it in one line or better to keep square brackets imo

File: src/rpcmisc.cpp [hunk 13]
```diff
@@ -172,6 +172,12 @@ UniValue mnsync(const UniValue& params, bool fHelp)
         return obj;
     }
 
+    if(strMode == "next")
+    {
+        masternodeSync.GetNextAsset();
+        return "sync updated to " + masternodeSync.GetAssetName();
```
Comment (UdjinM6) [inline src/rpcmisc.cpp]: Nice idea to make id more human readable :)
Probably a good idea to add `obj.push_back(Pair("RequestedMasternodeAssets", masternodeSync.GetAssetName()));` in `status` output too.

File: src/masternode-sync.cpp [hunk 14]
```diff
@@ -364,27 +390,26 @@ void CMasternodeSync::Process()
                 //     }
                 // }
 
-                //we'll start rejecting votes if we accidentally get set as synced too soon, this allows plenty of time
-                if(lastBudgetItem < GetTime() - MASTERNODE_SYNC_TIMEOUT){
-                    GetNextAsset();
+                // //we'll start rejecting votes if we accidentally get set as synced too soon, this allows plenty of time
+                // if(lastBudgetItem < GetTime() - MASTERNODE_SYNC_TIMEOUT){
```
Comment (UdjinM6) [inline src/masternode-sync.cpp]: why this was disabled?

File: src/main.cpp (hunk 12 above)
Comment (evan82) [inline src/main.cpp]: I thought the style guide for BTC was to have curly braces after the line break?

File: src/masternode-sync.cpp (hunk 14 above)
Comment (evan82) [inline src/masternode-sync.cpp]: accidental

File: src/main.cpp (hunk 12 above)
Comment (schinzelh) [inline src/main.cpp]: > Braces on new lines for namespaces, classes, functions, methods.
> Braces on the same line for everything else

https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

File: src/governance.cpp [hunk 15]
```diff
@@ -99,6 +97,116 @@ bool IsCollateralValid(uint256 nTxCollateralHash, uint256 nExpectedHash, std::st
...
+        }
+
+        CMasternode* pmn = mnodeman.Find(vote.vinMasternode);
+        if(pmn == NULL) {
+            LogPrint("mngovernance", "mngovernance - unknown masternode - vin: %s\n", vote.vinMasternode.ToString());
+            mnodeman.AskForMN(pfrom, vote.vinMasternode);
+            return;
+        }
+
+        mapSeenVotes.insert(make_pair(vote.GetHash(), SEEN_OBJECT_IS_VALID));
```
Comment (evan82) [inline src/governance.cpp]: hmm, no, but the code has a bug.  thx 👍

File: src/governance.cpp (hunk 7 above)
Comment (evan82) [inline src/governance.cpp]: Ah, yes. That was the bug I referenced above

File: src/governance.cpp (hunk 9 above)
Comment (evan82) [inline src/governance.cpp]: thanks, that's a good improvement

File: src/governance.cpp [hunk 16]
```diff
@@ -514,46 +494,43 @@ bool CGovernanceManager::UpdateGovernanceObject(CGovernanceVote& vote, CNode* pf
...
+    if(mapVotesByType.count(nTypeHash)) {
+        if(mapVotesByType[nTypeHash].nTime > vote.nTime){
+            strError = strprintf("new vote older than existing vote - %s", nTypeHash.ToString());
             LogPrint("mngovernance", "CGovernanceObject::AddOrUpdateVote - %s\n", strError);
             return false;
         }
-        if(vote.nTime - mapVotes[hash2].nTime < GOVERNANCE_UPDATE_MIN){
-            strError = strprintf("time between votes is too soon - %s - %lli", vote.GetHash().ToString(), vote.nTime - mapVotes[hash2].nTime);
+        if(vote.nTime - mapVotesByType[nTypeHash].nTime < GOVERNANCE_UPDATE_MIN){
+            strError = strprintf("time between votes is too soon - %s - %lli", nTypeHash.ToString(), vote.nTime - mapVotesByType[nHash].nTime);
```
Comment (UdjinM6) [inline src/governance.cpp]: typo `mapVotesByType[nHash]` -> `mapVotesByType[nTypeHash]` ?

Comment (UdjinM6) [inline src/governance.cpp]: in general: looks much cleaner now so after this ^^^ is addressed imo we are good to merge

Comment (evan82) [inline src/governance.cpp]: nope, that is correct, but upon further review there was a bug in GetTypeHash. I've added some comments to the code to make the intention clearer. We're trying to track when masternodes update their votes on governance objects. With 12.1 there's various types of votes (funding, valid, delete, etc), so this is the deterministic hash that will collide with the previous vote and allow the system to update.

Comment (UdjinM6) [inline src/governance.cpp]: Hmm.. ok, I missed the one in `GetTypeHash()` :) but shouldn't `mapVotesByType` be indexed via `nTypeHash` and not `nHash` here anyway?

Comment (evan82) [inline src/governance.cpp]: ha, yes. updated... I think we're good to go now?

**Comment by UdjinM6** [comment]
ACK squash merge 732a8a3 :)
@schinzelh

**Comment by crowning-** [comment]
@UdjinM6 @schinzelh : I noted "squash merge" in a couple of PRs recently. Does this mean I don't have to squash all my commits into one myself anymore?

**Comment by schinzelh** [comment]
@crowning- Yep, github introduced this wonderfull "squash and merge" option

image

**Pull request merged**